Fixes #11407 - Uppercase logins from LDAP break external user group sync

dLobatog opened the pull request:
When an user, say "DanielLobato" in LDAP gets an account in Foreman, the login
is automatically downcased to "daniellobato". This causes issues when external
user groups refresh, as we perform a 'where' query on User, using the logins
from LDAP to check what users belong in an external user group. So by calling
User.where { :name => ['DanielLobato'] } on usergroup.rb, we will get no
results and the user 'daniellobato' will be left unsynced.

To fix this we can downcase all logins coming from LDAP to compare them with
the ones in our database

dLobatog committed:
Fixes #11407 - Uppercase logins from LDAP break external user group sync

On LDAP the login can contain uppercase chars, for instance, "FOO". However
when we log in Foreman for the first time and have that account auto-created,
we can login using "foo". After that, our login will be saved as "foo" on
Foreman.

When a user group that contains said group is refreshed, we pull the names
from LDAP, auth_source.users_in_group(name). This will return an array
containing "FOO". After that, we will call usergroup.add_users(["FOO"])
which in turn calls User.where(:login => ["FOO"]). This will be empty since
our login in the database is "foo".

This commit fixes this issue in two places:
  One, by saving the login as it comes from LDAP (case aware), so that
  in the previous example 'FOO' would've been saved even if the user had
  try to login as 'foo'.

  Two, by making add_users and remove_users case insensitive.

domcleal commented:
a) "the login is automatically downcased" - where?  If in Foreman, that's a bug.  If it's in ldap_fluff, that's a bug, but I think I'm aware of it (in the generic member service).  It shouldn't be downcased, the original case should be kept all the way through.

b) User logins are case insensitive, but original case is preserved.  It sounds like .downcase would also be incorrect unless you specifically change something else to do User.where(:lower_login ..) rather than :login.

dLobatog commented:
@domcleal Sorry, I got this wrong. I've been digging deeper and the actual problem is this:
- On AD - the user can contain uppercase chars, for instance, "FOO". However when we log in Foreman for the first time and have that account auto-created, we can login using "foo". After that, our login will be saved as "foo" on Foreman.
- When a user group that contains said group is refreshed, we pull the names from LDAP, `auth_source.users_in_group(name)`. This will return an array containing "FOO". 
- After that, we will call `usergroup.add_users(["FOO"])` which in turn calls `User.where(:login => ["FOO"])`. This will be empty since our login in the database is "foo".

The current patch is hence not OK as if we had logged in as "FOO" in the previous example, the current patch would downcase the login coming from LDAP, and `User.where(:login => ["foo"])` will return empty. 

We could simply downcase the userlist in [add_users, usergroup.rb](https://github.com/theforeman/foreman/blob/develop/app/models/usergroup.rb#L64) and then search by lower_login. `remove_users` should be updated accordingly too.

domcleal commented:
> On AD - the user can contain uppercase chars, for instance, "FOO". However when we log in Foreman for the first time and have that account auto-created, we can login using "foo". After that, our login will be saved as "foo" on Foreman.

Yeah, that's incorrect because since 6be0508, it ought to be saved as "FOO" with lower_login as "foo".

> After that, we will call usergroup.add_users(["FOO"]) which in turn calls User.where(:login => ["FOO"]). This will be empty since our login in the database is "foo".

That should be calling `User.find_by_login("FOO")` which internally delegates to find_by_lower_login.

dLobatog commented:
I decided to fix it in both places, now if you try to login as 'foo', Foreman will still create your account as 'FOO' as that's what comes from LDAP. I've also made add_users and remove_users case insensitive.

domcleal commented on app/models/usergroup.rb:
```diff
@@ -62,11 +62,12 @@ def expire_topbar_cache(sweeper)
   def remove_users(userlist)
-    old_users = User.select { |user| userlist.include?(user.login) }
+    userlist.map(&:downcase!)
+    old_users = User.select { |user| userlist.include?(user.lower_login) }
```
Could this just use a .where like add_users?

domcleal commented on test/unit/user_test.rb:
```diff
@@ -138,10 +138,6 @@ def setup
     end
   end
 
-  test ".try_to_login if password is empty should return nil" do
```
Why remove this?

dLobatog replied:
There is a test right below this one, `when password is empty should return nil` that is identical.

domcleal commented on test/unit/user_test.rb:
```diff
@@ -617,27 +613,34 @@ def setup_user(operation)
     end
+
+    test "use LDAP login attribute as login" do
+      AuthSourceLdap.any_instance.stubs(:update_usergroups).
```
expects, not stubs?

domcleal commented on test/unit/usergroup_test.rb:
```diff
@@ -131,19 +131,20 @@ def populate_usergroups
   end
 
-  test 'add_users adds users in list and does not add nonexistent users' do
+  test 'add_users is case insensitive does not add nonexistent users' do
```
_and_ does not

dLobatog commented:
Thanks for the review, @domcleal . Updated according to your comments.

domcleal commented on test/unit/host_test.rb:
```diff
@@ -1366,6 +1366,7 @@ def teardown
       hosts = Host::Managed.search_for("facts.memory_mb > 6544")
       assert_equal hosts.count, 1
       assert_equal ["num001.example.com"], hosts.map { |h| h.name }.sort
+      binding.pry
```
Nope!

domcleal commented:
Working pretty well.  I hit a further issue when using the auto-create functionality, which calls the AuthSourceLdap#update_usergroups method updated in the PR, but with the lower case name.  You can see it here:

<pre>
2015-09-02T16:23:08 [sql] [D] Updating user groups for user upperldap
2015-09-02T16:23:08 [sql] [D]   User Load (0.1ms)  SELECT "users".* FROM "users" WHERE "users"."lower_login" = 'upperldap' ORDER BY firstname LIMIT 1
2015-09-02T16:23:08 [sql] [D]   Usergroup Load (0.1ms)  SELECT "usergroups".* FROM "usergroups" INNER JOIN "usergroup_members" ON "usergroups"."id" = "usergroup_members"."usergroup_id" WHERE "usergroup_members"."member_id" = 5
6 AND "usergroup_members"."member_type" = 'User' ORDER BY usergroups.name
2015-09-02T16:23:08 [sql] [D] String does not start with the prefix 'encrypted-', so AuthSourceLdap foreman.example.com was not decrypted
2015-09-02T16:23:08 [ldap] [D]   op search (2.6ms)  [ filter=, base= ]
2015-09-02T16:23:08 [ldap] [D]   op search (2.0ms)  [ filter=(uid=upperldap), base=dc=my-domain,dc=com ]
2015-09-02T16:23:08 [ldap] [D]   group_list (5.3ms)  [ user=upperldap ]
2015-09-02T16:23:08 [sql] [I] User 'UpperLDAP' auto-created from LDAP-foreman.example.com
</pre>

I think the User.try_to_auto_create_user method needs to pass user.login into this method, instead of login (which is the one the user logged in with).

dLobatog commented on test/unit/user_test.rb:
```diff
@@ -617,27 +613,34 @@ def setup_user(operation)
     test "enabled on-the-fly registration" do
       AuthSourceLdap.any_instance.expects(:update_usergroups).
-                                  with('fakeuser').returns(true)
+                                  with('FoOBaR').returns(true)
```
This `FoOBaR` and the one in line 640 test that we're calling update_usergroups on `try_to_auto_create`.

dLobatog commented:
@domcleal Thanks for testing. I've updated the PR and left a comment to indicate where I'm testing that behavior, now I'm calling update_usergroups with user.login to preserve casing.

domcleal commented:
Seems the same issue happens on logins where the user _isn't_ automatically created - there's a second update_usergroups call in the User.try_to_login method that's again using the (user-provided) login rather than user.login.

dLobatog commented:
@domcleal I've updated the call in try_to_login, but I think at that point the function, `login` could only be the same as `user.login`, because `user` comes from `User.find_by_login(login)`. Rebased.

domcleal commented:
`login` could be lower case from user input, but find_by_login will find case insensitively, so it could be returning an existing user with upper/mixed case logins, so user.login ought to be better than plain login.

domcleal commented:
Merged as 2ef6f4d, thanks @eLobato.
